common/argparse.cpp

```diff
@@ -0,0 +1,1362 @@
+  if (option.kind == Arg::Kind::Flag) {
+    out_ << "--" << (option.default_flag ? "no-" : "") << option.info.name;
+    return;
+  }
+  out_ << "--" << option.info.name;
+  if (option.kind != Arg::Kind::MetaActionOnly) {
+    out_ << (option.has_default ? "[" : "") << "=";
+    PrintArgValueUsage(option);
+    if (option.has_default) {
+      out_ << "]";
+    }
+  }
+}
+
+void Args::MetaPrinter::PrintShortName(const Arg& arg) const {
+  CARBON_CHECK(!arg.info.short_name.empty()) << "No short name to use.";
+  out_ << "-" << arg.info.short_name;
+}
+
+void Args::MetaPrinter::PrintArgShortValues(const Arg& arg) const {
+  CARBON_CHECK(arg.kind == Arg::Kind::OneOf)
+      << "Only one-of arguments have interesting value snippets to print.";
+  bool first = true;
+  for (llvm::StringRef value_string : arg.value_strings) {
+    if (first) {
+      first = false;
+    } else {
+      out_ << ", ";
+    }
+    out_ << value_string;
+  }
+}
+void Args::MetaPrinter::PrintArgLongValues(const Arg& arg,
+                                           llvm::StringRef indent) const {
+  out_ << indent << "Possible values:\n";
+  // TODO: It would be good to add help text for each value and then print it
+  // here.
+  for (int i = 0, size = arg.value_strings.size(); i < size; ++i) {
+    llvm::StringRef value_string = arg.value_strings[i];
+    out_ << indent << "- " << value_string;
+    if (arg.has_default && i == arg.default_value_index) {
+      out_ << " (default)";
+    }
+    out_ << "\n";
+  }
+}
+
+void Args::MetaPrinter::PrintArgHelp(const Arg& arg,
+                                     llvm::StringRef indent) const {
+  // Print out the main help text.
+  PrintTextBlock(indent, arg.info.help);
+
+  // Then print out any help based on the values.
+  switch (arg.kind) {
+    case Arg::Kind::Integer:
+      if (arg.has_default) {
+        out_ << "\n";
+        out_ << indent << "Default value: " << arg.default_integer << "\n";
+      }
+      break;
+    case Arg::Kind::String:
+      if (arg.has_default) {
+        out_ << "\n";
+        out_ << indent << "Default value: " << arg.default_string << "\n";
+      }
+      break;
+    case Arg::Kind::OneOf:
+      out_ << "\n";
+      PrintArgLongValues(arg, indent);
+      break;
+    case Arg::Kind::Flag:
+    case Arg::Kind::MetaActionOnly:
+      // No value help.
+      break;
+    case Arg::Kind::Invalid:
+      CARBON_FATAL() << "Argument configured without any action or kind!";
+  }
+}
+
+void Args::MetaPrinter::PrintRawUsageCommandAndOptions(
+    const Command& command, int max_option_width) const {
+  // Recursively print parent usage first with a compressed width.
+  if (command.parent) {
+    PrintRawUsageCommandAndOptions(*command.parent, MaxParentOptionUsageWidth);
+    out_ << " ";
+  }
+
+  out_ << command.info.name;
+
+  // Buffer the options rendering so we can limit its length.
+  std::string buffer_str;
+  llvm::raw_string_ostream buffer_out(buffer_str);
```

Comment:
Here and the other spot you're using raw_string_ostream, consider instead using llvm::buffer_ostream. I think it's a mild simplification for your use-case, since you wouldn't need to separately declare the buffer, just use `buffer.str()` (since it inherits from raw_svector_ostream).

(I've spent a lot of time looking at the stream apis now)